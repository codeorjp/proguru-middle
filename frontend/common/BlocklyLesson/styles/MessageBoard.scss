@import 'common/styles/variables.scss';
.messageboard {
  width: 100%;
  height: 100%;
  position: relative;
  outline: none;

  .messageboardWindow {
    margin: 0 auto;
    height: 100%;

    .messageboardHeader {
      border-radius: 10px 10px 0 0;
    }

    .messageboardBody {
      background: #fff;
      padding: 20px;
    
      .messageboardBodyHeader {
        margin-bottom: 1px;
      }

      .messageboardFooterContents {
        width: 90%;
        margin: 0 auto;
        padding-top: 15px;
        padding-bottom: 15px;
      }

      .messageboardName {
        text-align: center;
        padding-bottom: 15px;
        font-size: 1.4rem;
      }

      .messageboardContents {
        padding: 10px 10px;
        height: 400px;
        overflow-y: scroll;
        overflow-x: hidden;
      }

      .messageboardMoreButton {
        background: $blue-1;
        border: 0;
        padding: 10px 20px;
        border-radius: 100px;
        color: #fff;
      }
    }
  }
}

.sendMessage {
  text-align: right;
  margin: 10px 0;
}

.receiveMessage {
  text-align: left;
  margin: 10px 0;
}

.overlay {
  border-radius: 1px;
  content: "";
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  width: 100%;
  height: 100%;
  z-index: 100;
  color: white;
  background: rgba(#000, 0.7);
  display: flex;
  justify-content: center;
  align-items: center;
  text-align: center;

  p {
    font-size: 22px;
  }
}

.displayNone {
  display: none;
}
