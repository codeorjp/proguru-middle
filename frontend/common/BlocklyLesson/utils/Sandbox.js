import _ from "lodash";

class Sandbox {
  constructor(interpreter) {
    this.interpreter = interpreter;
    this.code = "";
  }

  init(store, blockFunctions, code) {
    const blockApis = _.map(blockFunctions, (func) => func(store));
    this.code = code;
    this.interpreter.init(blockApis);
  }

  async eval() {
    await this.interpreter.eval(this.code);
  }

  reset() {
    this.interpreter.reset();
  }
}

export default Sandbox;
