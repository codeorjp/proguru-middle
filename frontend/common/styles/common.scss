@import 'common/styles/messageBoard';

// normalize入れておく
.lessonCommon {
  @import url(https://cdnjs.cloudflare.com/ajax/libs/normalize/3.0.3/normalize.min.css);
  @import './variables.scss';
  // base style
  font-size: 14px;
  font-weight: bold;
  background: $lesson-bg;
  display: flex;
  flex-direction: column;
  flex: 1 1 auto;

  * {
    box-sizing: border-box;
  }

  a,
  a:visited,
  a:hover {
    color: #333;
    background: none;
  }

  p {
    margin: 0;
  }
}
