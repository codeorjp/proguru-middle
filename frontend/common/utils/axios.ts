import axios from "axios";
import { csrfToken } from "rails-ujs";

axios.defaults.headers.common["X-CSRF-TOKEN"] = csrfToken();
axios.defaults.headers.common["X-Requested-With"] = "XMLHttpRequest";
axios.defaults.headers.common["Content-Type"] = "application/json";

export default axios;
