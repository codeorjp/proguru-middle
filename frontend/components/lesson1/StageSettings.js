import Stage1 from "./Stage1";
import Stage2 from "./Stage2";
import Stage3 from "./Stage3";
import Stage4 from "./Stage4";
import Stage5 from "./Stage5";
import Stage6 from "./Stage6";
import Stage7 from "./Stage7";
import Stage8 from "./Stage8";

const StageSettings = {
  1: Stage1,
  2: Stage2,
  3: Stage3,
  4: Stage4,
  5: Stage5,
  6: Stage6,
  7: Stage7,
  8: Stage8,
};

export default StageSettings;
