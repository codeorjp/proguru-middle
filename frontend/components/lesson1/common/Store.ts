import { action, decorate, observable, toJS } from "mobx";
import Storage from "components/lesson1/common/util/Storage";
import { DesignSchemaType } from "common/types/DesignSchemaType";
import { ComponentStyleType } from "common/types/ComponentStyleType";
import { LessonType } from "common/types/LessonType";
import { StageType } from "common/types/StageType";

type ClickElement = {
  title: string;
  className: string;
  style: ComponentStyleType;
  editDesignComponent: string[];
};

type StageConfig = {
  className: string;
  editDesignComponent: string[];
}[];

class Store {
  storage: Storage;
  schemas: DesignSchemaType;
  clickElm: ClickElement;
  stageConfig: StageConfig;
  sendSpeechText: string;
  receiveSpeechText: string;
  isFirstTime: boolean;
  isStageFirstTime: boolean;

  constructor(
    public userId: number | string,
    public lesson: LessonType,
    public stage: StageType,
    public submittedWorkspace: any
  ) {
    this.storage = new Storage(`${window.location.href}-user${this.userId}`);
    this.schemas = {
      SendSpeechBalloon: {
        style: {
          font: { fontSize: "16px", textAlign: "right" },
          color: "#000",
          border: {},
          background: "",
          borderRadius: "",
        },
      },
      SendProfileIcon: {
        style: {
          imgSize: {
            width: "200px",
            height: "200px",
          },
          border: {
            borderWidth: "1px",
          },
          borderRadius: "",
          background: "#fff",
        },
      },
      ReceiveSpeechBalloon: {
        style: {
          font: { fontSize: "16px", textAlign: "right" },
          color: "#000",
          border: {},
          borderRadius: "",
          background: "",
        },
      },
      ReceiveProfileIcon: {
        style: {
          imgSize: {
            width: "200px",
            height: "200px",
          },
          border: {
            borderWidth: "1px",
          },
          borderRadius: "",
          background: "#fff",
        },
      },
      ChatInput: {
        style: {
          font: { fontSize: "14px" },
          color: "#000",
          border: {},
          borderRadius: "",
          background: "#fff",
          padding: "",
        },
      },
      SubmitButton: {
        style: {
          font: { fontSize: "14px" },
          color: "#000",
          border: {},
          borderRadius: "",
          background: "#fff",
          padding: "",
        },
      },
      ChatBody: {
        style: {
          background: "#fff",
        },
      },
    };
    this.clickElm = {
      title: "", // 編集するコンポーネントタイトル
      className: "", // CSSクラス名
      style: {}, // 当てるスタイル
      editDesignComponent: [], // 編集コンポーネント
    };
    this.stageConfig = [
      {
        // アタッチするデザイン編集項目
        className: "", // CSSクラス名
        editDesignComponent: [], // 編集コンポーネント
      },
    ];
    this.sendSpeechText = "こんにちは，私の名前は美咲です。";
    this.receiveSpeechText = "こんにちは，私の名前は翔太です。";
    this.isFirstTime = this.checkFirstTime(
      `lesson${this.lesson.number}:isFirstTime`
    );
    this.isStageFirstTime = this.checkFirstTime(
      `lesson${this.lesson.number}-stage${this.stage.number}`
    );

    // sessionStorageのrestore処理
    this.schemas = this.storage.restoreWorkspace(this.schemas);
  }

  restoreSubmittedWorkspace(schemas: DesignSchemaType): void {
    if (schemas) {
      this.schemas = schemas;
    }
  }

  checkFirstTime(url: string): boolean {
    try {
      const isFirstTime = window.localStorage.getItem(url);
      return isFirstTime !== "false";
    } catch {
      return true;
    }
  }

  setFirstTimeDone(url: string): void {
    this.isFirstTime = false;
    try {
      window.localStorage.setItem(url, "false");
    } catch {}
  }

  setClickElm(clickElm: ClickElement): void {
    this.clickElm = clickElm;
  }

  setStyle(className: string, style: ComponentStyleType): void {
    const prevStyle = toJS(this.schemas[className].style);
    const nextStyle = {
      ...prevStyle,
      ...style,
    };
    this.schemas[className].style = nextStyle;
    this.clickElm.style = nextStyle;

    this.storage.saveWorkspace(this.schemas);
  }

  setSendSpeechText(text: string): void {
    this.sendSpeechText = text;
  }

  setReceiveSpeechText(text: string): void {
    this.receiveSpeechText = text;
  }

  setStageConfig(config: StageConfig): void {
    this.stageConfig = config;
  }
}

decorate(Store, {
  clickElm: observable,
  schemas: observable,
  isFirstTime: observable,
  sendSpeechText: observable,
  receiveSpeechText: observable,
  setClickElm: action,
  setStageConfig: action,
  setStyle: action,
  setSendSpeechText: action,
  setReceiveSpeechText: action,
  setFirstTimeDone: action,
});

export default Store;
