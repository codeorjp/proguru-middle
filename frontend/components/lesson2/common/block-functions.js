export const display = (store) => ({
  name: "display",
  def: (varName, blockId) => {
    store.display(varName, blockId);
  },
});

export const assignText = (store) => ({
  name: "assignText",
  def: (text) => {
    store.assignText(text);
    return "text";
  },
});

export const getMeridiem = (store) => ({
  name: "getMeridiem",
  def: () => store.getMeridiem(),
});

export const connect = (store) => ({
  name: "connect",
  def: (varName, callback, blockId) =>
    new Promise((resolve) => {
      store.connect(varName, callback, blockId, resolve);
    }),
});

export const send = (store) => ({
  name: "send",
  def: (varName, blockId) =>
    new Promise((resolve) => {
      store.send(varName, blockId, resolve);
    }),
});

export const receiveResponse = (store) => ({
  name: "receiveResponse",
  def: (callback, blockId) => {
    store.receiveResponse(callback, blockId);
  },
});

export const highlightBlock = (store) => ({
  name: "highlightBlock",
  def: (blockId) => {
    store.skip(blockId);
  },
});
