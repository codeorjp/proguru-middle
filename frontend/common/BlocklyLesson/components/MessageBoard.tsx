import _ from "lodash";
import React from "react";
import { inject, observer } from "mobx-react";
import { toJS } from "mobx";
import flattenObject from "common/utils/flattenObject";
import SubmitArea from "common/BlocklyLesson/components/SubmitArea";
import styles from "common/BlocklyLesson/styles/MessageBoard.scss";
import ReadmoreButton from "common/BlocklyLesson/components/message_board/ReadmoreButton";
import Timeline from "common/BlocklyLesson/components/message_board/Timeline";
import ExecuteOverlay from "common/BlocklyLesson/components/message_board/ExecuteOverlay";

interface MessageBoardProps {
  store?: any;
  types: string[];
}

class MessageBoard extends React.Component<MessageBoardProps> {
  styleNames = [
    "SendSpeechBalloon",
    "SendProfileIcon",
    "ReceiveSpeechBalloon",
    "ReceiveProfileIcon",
    "ChatInput",
    "SubmitButton",
    "ChatBody",
  ];
  styles: {
    [key: string]: object;
  };

  constructor(props) {
    super(props);
    this.styles = {};
    this.getPastMessages = this.getPastMessages.bind(this);
  }

  getPastMessages() {
    this.props.store.getPastMessages();
  }

  render() {
    const { types } = this.props;
    const {
      messages,
      messageBoardStyle,
      isExecute,
      getPastMessages,
      existPastMessage,
      isShowExecuteOverlay,
    } = this.props.store;

    if (messageBoardStyle) {
      // レッスンで提出済みのスタイルにここで上書きされる
      this.styleNames.forEach((styleName) => {
        this.styles[styleName] = flattenObject(
          toJS(messageBoardStyle)[styleName].style
        );
      });
    }

    const sortedUniqueMessages = _.sortBy(_.uniqBy(messages, "messageId"), [
      "createdAt",
    ]);
    const {
      messageboard,
      messageboardWindow,
      messageboardBody,
      messageboardBodyHeader,
      messageboardName,
      messageboardContents,
      messageboardBodyFooter,
      messageboardMoreButton,
    } = styles;

    return (
      <div>
        <div className={messageboard}>
          <div className={messageboardWindow}>
            <div
              className={messageboardBody}
              style={messageBoardStyle ? this.styles.ChatBody : null}
            >
              <div className={messageboardBodyHeader}>
                <div className={messageboardName}>
                  <p>クラスのチャット</p>
                </div>
              </div>
              <div className={`${messageboardContents} messageBoardContents`}>
                <ReadmoreButton
                  isShow={!!getPastMessages}
                  handleClick={this.getPastMessages}
                  className={messageboardMoreButton}
                  isExecute={isExecute}
                  hasPastMessage={existPastMessage}
                />
                <Timeline
                  messages={sortedUniqueMessages}
                  userId={this.props.store.userId}
                  hasStudentsDesign={!!messageBoardStyle}
                  styles={this.styles}
                />
              </div>
              <div className={messageboardBodyFooter}>
                <SubmitArea types={types} />
              </div>
            </div>
          </div>
          <ExecuteOverlay
            isShow={isShowExecuteOverlay}
            isExecute={isExecute}
            styles={styles}
          />
        </div>
      </div>
    );
  }
}

export default inject("store")(observer(MessageBoard));
