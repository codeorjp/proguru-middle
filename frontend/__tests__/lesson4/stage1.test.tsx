/* eslint-disable no-undef */
import _ from "lodash";
import axios from "axios";
import moment from "moment";
import React from "react";
import Blockly from "blockly";
import * as B from "blockly/core";
import BlocklyJS from "blockly/javascript";
import { Provider } from "mobx-react";
import renderer from "react-test-renderer";
import Store from "components/lesson4/common/Store";
import Sandbox from "common/BlocklyLesson/utils/Sandbox";
import MessageBoard from "common/BlocklyLesson/components/MessageBoard";
import * as BlockFunctions from "components/lesson4/common/block-functions";
import StageSettings from "components/lesson4/StageSettings";
import { NodeEval, assignBlocks } from "./utils";

jest.mock("axios");
const mockedAxios = axios as jest.Mocked<typeof axios>;

let cBlockly: any;
let store: any;
let interpreter: any;
let sandbox: any;
let workspace: Blockly.Workspace;
let component: any;

beforeEach(() => {
  // axios
  mockedAxios.get.mockResolvedValue({ data: {} });

  // Storeのインスタンスを生成
  // constructorのaxios.get()でError検出する
  store = new Store(
    0, // userId
    "生徒1", // nickName
    0, // studnetNumber
    null, // userIcon
    { number: 4 }, // lesson
    { number: 1 }, // stage
    null // submittedWorkspace
  );
  store.setSampleImage();

  // Blockly
  cBlockly = _.clone(B);
  assignBlocks(
    cBlockly,
    BlocklyJS,
    StageSettings[store.stage.number].workspace.tools
  );
  assignBlocks(
    cBlockly,
    BlocklyJS,
    StageSettings[store.stage.number].workspace.unDeletableBlocks
  );
  workspace = new Blockly.Workspace();

  // Sandbox
  interpreter = new NodeEval();
  sandbox = new Sandbox(interpreter);

  // Component
  component = renderer.create(
    <Provider store={store}>
      <MessageBoard types={["Image"]} />
    </Provider>
  );
});

const editCreatedAt = () => {
  const { messages } = store;
  messages.forEach((_message: object, index: number) => {
    messages[index].createdAt = moment("20200202", "YYYYMMDD");
  });
};

describe("Lesson4 Stage1", () => {
  test("build blocks -> eval code -> display content(sampleImage 1)", () => {
    const onExecute = workspace.newBlock("onExecute");
    const display = workspace.newBlock("display");
    const sampleImage = workspace.newBlock("sampleImage");

    onExecute.getInput("DO").connection.connect(display.previousConnection);
    display
      .getInput("VARNAME")
      .connection.connect(sampleImage.outputConnection);

    const code = BlocklyJS.workspaceToCode(workspace);
    sandbox.init(store, BlockFunctions, code);
    sandbox.eval();

    editCreatedAt();

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("build blocks -> eval code -> display content(sampleImage 1, 2)", () => {
    const onExecute = workspace.newBlock("onExecute");
    const display1 = workspace.newBlock("display");
    const display2 = workspace.newBlock("display");
    const sampleImage1 = workspace.newBlock("sampleImage");
    const sampleImage2 = workspace.newBlock("sampleImage");

    onExecute.getInput("DO").connection.connect(display1.previousConnection);
    display1
      .getInput("VARNAME")
      .connection.connect(sampleImage1.outputConnection);
    display1.nextConnection.connect(display2.previousConnection);
    display2
      .getInput("VARNAME")
      .connection.connect(sampleImage2.outputConnection);
    sampleImage2.setFieldValue("2", "NUMBER");

    const code = BlocklyJS.workspaceToCode(workspace);
    sandbox.init(store, BlockFunctions, code);
    sandbox.eval();

    editCreatedAt();

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
