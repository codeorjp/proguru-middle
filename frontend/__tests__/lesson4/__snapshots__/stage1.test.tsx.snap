// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lesson4 Stage1 build blocks -> eval code -> display content(sampleImage 1) 1`] = `
<div>
  <div>
    <div>
      <div
        style={null}
      >
        <div>
          <div>
            <p>
              クラスのチャット
            </p>
          </div>
        </div>
        <div
          className="undefined messageBoardContents"
        >
          <div>
            <div
              className="sendMessageArea"
            >
              <div>
                <p>
                  <span>
                    12:00 AM
                  </span>
                  <span>
                    出席番号：0
                  </span>
                  <span>
                    生徒1
                  </span>
                </p>
                <div
                  aria-describedby="popup-1"
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                  style={
                    Object {
                      "height": "0px",
                    }
                  }
                >
                  <img
                    alt=""
                    src="test-file-stub"
                  />
                </div>
              </div>
              <div
                style={null}
              >
                <img
                  alt=""
                  src={null}
                />
              </div>
            </div>
          </div>
        </div>
        <div>
          <div>
            <div
              style={
                Object {
                  "display": "block",
                }
              }
            >
              <button
                onClick={[Function]}
                type="button"
              >
                <input
                  accept="image/jpeg"
                  onChange={[Function]}
                  onClick={[Function]}
                  style={
                    Object {
                      "display": "none",
                    }
                  }
                  type="file"
                />
                <i
                  className="bi bi-image undefined"
                  style={
                    Object {
                      "cursor": "pointer",
                    }
                  }
                />
              </button>
            </div>
            <div>
              <input
                disabled={true}
                onChange={[Function]}
                style={null}
                value=""
              />
            </div>
            <div>
              <button
                disabled={true}
                onClick={[Function]}
                style={null}
                type="button"
              >
                送信する
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div>
      <p>
        プログラムを組んだら
        <br />
        「実行する」ボタンを押してね
      </p>
    </div>
  </div>
</div>
`;

exports[`Lesson4 Stage1 build blocks -> eval code -> display content(sampleImage 1, 2) 1`] = `
<div>
  <div>
    <div>
      <div
        style={null}
      >
        <div>
          <div>
            <p>
              クラスのチャット
            </p>
          </div>
        </div>
        <div
          className="undefined messageBoardContents"
        >
          <div>
            <div
              className="sendMessageArea"
            >
              <div>
                <p>
                  <span>
                    12:00 AM
                  </span>
                  <span>
                    出席番号：0
                  </span>
                  <span>
                    生徒1
                  </span>
                </p>
                <div
                  aria-describedby="popup-3"
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                  style={
                    Object {
                      "height": "0px",
                    }
                  }
                >
                  <img
                    alt=""
                    src="test-file-stub"
                  />
                </div>
              </div>
              <div
                style={null}
              >
                <img
                  alt=""
                  src={null}
                />
              </div>
            </div>
          </div>
          <div>
            <div
              className="sendMessageArea"
            >
              <div>
                <p>
                  <span>
                    12:00 AM
                  </span>
                  <span>
                    出席番号：0
                  </span>
                  <span>
                    生徒1
                  </span>
                </p>
                <div
                  aria-describedby="popup-5"
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                  style={
                    Object {
                      "height": "0px",
                    }
                  }
                >
                  <img
                    alt=""
                    src="test-file-stub"
                  />
                </div>
              </div>
              <div
                style={null}
              >
                <img
                  alt=""
                  src={null}
                />
              </div>
            </div>
          </div>
        </div>
        <div>
          <div>
            <div
              style={
                Object {
                  "display": "block",
                }
              }
            >
              <button
                onClick={[Function]}
                type="button"
              >
                <input
                  accept="image/jpeg"
                  onChange={[Function]}
                  onClick={[Function]}
                  style={
                    Object {
                      "display": "none",
                    }
                  }
                  type="file"
                />
                <i
                  className="bi bi-image undefined"
                  style={
                    Object {
                      "cursor": "pointer",
                    }
                  }
                />
              </button>
            </div>
            <div>
              <input
                disabled={true}
                onChange={[Function]}
                style={null}
                value=""
              />
            </div>
            <div>
              <button
                disabled={true}
                onClick={[Function]}
                style={null}
                type="button"
              >
                送信する
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div>
      <p>
        プログラムを組んだら
        <br />
        「実行する」ボタンを押してね
      </p>
    </div>
  </div>
</div>
`;
